### PR TITLE
Documentation and feature flag

### PR DESCRIPTION
**Pull request by jens-esr**
Added the ability to switch between thread local and a split socket via a feature flag.

An assert was added in the builder if only one of the two flags is set. 
Is there a way to check this at compile time?

Documentation was added. 

Currently there is still a compiler warning that UnixDatagram is not used when compiling for a shared socket.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,67 @@
...
+## Android
+In case logd-logger is compiled for Android, 
+the log messages are sent to Android's logd socket.
+
+## Other OS
+For other Operating Systems the logging messages are sent to standard out.
+These are displayed in logd format.
+
+# Usage
+
```
Comment (flxo) [inline README.md]: Inconsistent newlines after the section.

File: Cargo.toml [hunk 2]
```diff
@@ -18,6 +18,7 @@ keywords = [
 env_logger = { version = "0.8.2", features = ["regex"], default-features = false }
 log = { version = "0.4.11", features = ["std"] }
 time = "0.2.23"
+lazy_static = { version = "1.4.0", optional = true }
```
Comment (flxo) [inline Cargo.toml]: The deps should be in alphanumeric order.

File: Cargo.toml [hunk 3]
```diff
@@ -30,3 +31,8 @@ winapi = { version = "0.3.9", features = ["processthreadsapi"] }
 
 [target.'cfg(target_os = "redox")'.dependencies]
 redox_syscall = "0.2.1"
+
+[features]
+default = ["locked"]
+tls = []
+locked = ["lazy_static"]
```
Comment (flxo) [inline Cargo.toml]: `locked` is probably not the right name. With this  feature one socket is shared. Rename this feature to `shared`.

File: README.md [hunk 4]
```diff
@@ -1,4 +1,67 @@
...
+Logd-logger is configured once when initiating.
+Messages are sent using the macros `error!`, `warn!`, `info!`, `debug!` and `trace!`.
+
+## Android
+In case logd-logger is compiled for Android, 
+the log messages are sent to Android's logd socket.
+
+## Other OS
+For other Operating Systems the logging messages are sent to standard out.
+These are displayed in logd format.
```
Comment (flxo) [inline README.md]: I'd replace `logd format` with `logcat` and add a link.

File: Cargo.toml [hunk 5]
```diff
@@ -18,6 +18,7 @@ keywords = [
 env_logger = { version = "0.8.2", features = ["regex"], default-features = false }
 log = { version = "0.4.11", features = ["std"] }
 time = "0.2.23"
+lazy_static = { version = "1.4.0", optional = true }
 
 [target.'cfg(target_os = "android")'.dependencies]
 bytes = "1.0.0"
```
Comment (flxo) [inline Cargo.toml]: We should change this to `bytes = "1"`

File: README.md [hunk 6]
```diff
@@ -1,4 +1,67 @@
...
+
+```toml
+[dependencies]
+logd-logger = "0.1.0"
+```
+
+Next:
+In the example the logging is initialized. First a [`Builder`] is created.
+Then the log level is set. All messages with lower log level are dropped.
+Then a tag is set which is added in front of each log message.
```
Comment (flxo) [inline README.md:37]: This is grammatically wrong ;-)

File: README.md [hunk 7]
```diff
@@ -1,4 +1,67 @@
...
+Next, the module name is set to be prepended.
+After the configuration, it is tried to initialize the global logger.
+
+Afterwards five different macros can be used for logging.
```
Comment (flxo) [inline README.md]: This can be omitted since everyone who uses this crates know the Rust logger concept.

File: src/lib.rs [hunk 8]
```diff
@@ -1,10 +1,93 @@
...
+     static SOCKET: UnixDatagram = {
+        let socket = std::os::unix::net::UnixDatagram::unbound().expect("Failed to create socket");
+        socket.connect("/dev/socket/logdw").expect("Failed to connect to /dev/socket/logdw");
+        socket
+    };
+}
+#[cfg(all(feature = "locked", not(feature = "tls"), target_os = "android"))]
+lazy_static! {
+    static ref SOCKET: std::os::unix::net::UnixDatagram = {
+        let socket = std::os::unix::net::UnixDatagram::unbound().expect("Failed to create socket");
```
Comment (flxo) [inline src/lib.rs]: Add a `use std::os::unix::net::UnixDatagram`. It's used twice and reads bulky.

File: src/lib.rs [hunk 9]
```diff
@@ -1,10 +1,93 @@
...
+        socket
+            .connect("/dev/socket/logdw")
```
Comment (flxo) [inline src/lib.rs]: Refactor the path to the socket into a constant.

File: src/lib.rs [hunk 10]
```diff
@@ -71,16 +154,23 @@ impl From<log::Level> for Priority {
     }
 }
 
-/// Log buffer ids
+/// Different Log Buffer for Android
```
Comment (flxo) [inline src/lib.rs]: It's still `Log buffer ids` This is the term that is used within Android.

File: src/lib.rs [hunk 11]
```diff
@@ -379,16 +464,24 @@ impl Log for Logger {
             buffer.put(message.as_bytes());
             buffer.put_u8(0);
 
-            SOCKET.with(|f| f.send(&buffer).expect("Logd socket error"));
+            #[cfg(all(feature = "tls", not(feature = "locked"), target_os = "android"))]
+            {
```
Comment (flxo) [inline src/lib.rs]: Think this bracket is not needed.

File: src/lib.rs [hunk 12]
```diff
@@ -379,16 +464,24 @@ impl Log for Logger {
...
+                SOCKET.with(|f| f.send(&buffer).expect("Logd socket error"));
+            }
+
+            #[cfg(all(feature = "locked", not(feature = "tls"), target_os = "android"))]
+            {
```
Comment (flxo) [inline src/lib.rs]: Think this bracket is not needed.

File: src/lib.rs [hunk 13]
```diff
@@ -1,10 +1,93 @@
+//! `logd-logger`
```
Comment (flxo) [inline src/lib.rs]: Theres a crate that generates the Readme out of `lib.rs`. Can we use that in order to not maintain the same doc twice?

**Review by flxo — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 13 above)
Comment (jens-esr) [reply src/lib.rs]: There is a cargo plugin that does the trick. Should we use that? [cargo-sync-readme](https://crates.io/crates/cargo-sync-readme)

Comment (flxo) [reply src/lib.rs]: Sounds like a plan. Go for it!

File: README.md (hunk 6 above)
Comment (jens-esr) [reply README.md:37]: I don't get what is wrong.

**Pull request merged**